ops/model.py

```diff
@@ -606,40 +674,161 @@ def open_port(self, protocol: typing.Literal['tcp', 'udp', 'icmp'],
         last unit that opened that port has closed it. However, this is
         usually not an issue; normally charms should make the same
         close_port() call from every unit.
 
+        Use :func:`set_ports()` for a more declarative approach where all
+        of the ports that should be open are provided in a single call.
+        For example, ``set_ports()`` will close all open ports.
+
         Args:
             protocol: String representing the protocol; must be one of
                 'tcp', 'udp', or 'icmp' (lowercase is recommended, but
                 uppercase is also supported).
             port: The port to open. Required for TCP and UDP; not allowed
                 for ICMP.
-        """
-        self._backend.close_port(protocol.lower(), port)
 
-    def opened_ports(self) -> Set['OpenedPort']:
+        Raises:
+            ModelError: If ``port`` is provided when ``protocol`` is 'icmp'
+                or ``port`` is not provided when ``protocol`` is 'tcp' or
+                'udp'.
+        """
+        normalised_protocol : str = protocol.lower()
+        if normalised_protocol == 'icmp' and port is not None:
+            raise ModelError("icmp cannot have a port number specified")
+        elif normalised_protocol in ('tcp', 'udp') and port is None:
+            raise ModelError(f"{normalised_protocol} must have a port number specified")
+        self._backend.close_port(normalised_protocol, port)
+
+    def opened_ports(self) -> Set['Port']:
         """Return a list of opened ports for this unit."""
         return self._backend.opened_ports()
 
+    def set_ports(self, *ports: Union[int, 'Port']) -> None:
+        """Set the open ports for this unit, closing any others that are open.
+
+        Register intent with Juju that the application should be accessed on
+        the given ports and no others, but the ports are not actually opened
+        externally until the admin runs "juju expose" and any closed ports are
+        not actually closed externally until the admin runs "juju unexpose".
+
+        On Kubernetes sidecar charms, the ports opened are not strictly
+        per-unit: Juju will open the union of ports from all units.
+        However, normally charms should make the same set_ports() call from
+        every unit.
+
+        Use :func:`open_port()` and :func:`close_port()` to manage ports
+        individually.
+
+        Args:
+            ports: The ports to open. Provide an int to open a TCP port, or
+                a :class:`Port` to open a port for another protocol.
+        """
+        # Normalise to get easier comparisons.
+        existing_ports = {
+            (port.protocol, port.port)
+            for port in self._backend.opened_ports()
+        }
+        desired_ports = {
+            ('tcp', port) if isinstance(port, int) else (port.protocol, port.port)
+            for port in ports
+        }
+        for protocol, port in existing_ports.difference(desired_ports):
```

Comment:
Do you think `in existing_ports - desired_ports` is clearer (and similar below)? I kinda think it is, don't have strong feelings though.